File: src/main/java/com/bajdcc/LALR1/interpret/module/net/ModuleNetServerHandler.java
package com.bajdcc.LALR1.interpret.module.net;

import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import io.netty.util.ReferenceCountUtil;

import java.util.Queue;

import static com.bajdcc.LALR1.interpret.module.net.ModuleNetServer.CHANNEL_GROUP;

/**
 * 【模块】通讯服务端处理
 *
 * @author bajdcc
 */
public class ModuleNetServerHandler extends ChannelInboundHandlerAdapter {

	private Queue<String> msgQueue;

	public ModuleNetServerHandler(Queue<String> msgQueue) {
		this.msgQueue = msgQueue;
	}

	@Override
	public void channelActive(ChannelHandlerContext ctx) {
		Channel ch = ctx.channel();
		CHANNEL_GROUP.add(ch);
		if (CHANNEL_GROUP.size() > 0) {
			CHANNEL_GROUP.writeAndFlush(
					String.format("{ \"origin\": \"%s\", \"addr\": \"%s\", \"type\": \"INFO\", \"content\": \"Hello, client!\" }\r\n",
							ch.remoteAddress().toString(), ch.localAddress().toString()));
		}
	}

	@Override
	public void channelInactive(ChannelHandlerContext ctx) {
		Channel ch = ctx.channel();
		msgQueue.add(String.valueOf(String.format("{ \"addr\": \"%s\", \"type\": \"INFO\", \"content\": \"Bye, client!\" }\r\n",
				ch.remoteAddress().toString())));
	}

	@Override
	public void channelRead(ChannelHandlerContext ctx, Object msg) {
		msgQueue.add(String.valueOf(msg));
		ReferenceCountUtil.release(msg);
	}

	@Override
	public void channelReadComplete(ChannelHandlerContext ctx) {
		ctx.flush();
	}

	@Override
	public void handlerRemoved(ChannelHandlerContext ctx) {
		ctx.close();
	}

	@Override
	public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
		cause.printStackTrace();
		CHANNEL_GROUP.remove(ctx.channel());
		ctx.close();
	}
}

File: src/main/kotlin/com/bajdcc/LALR1/interpret/module/net/ModuleNetServerHandler.kt
package com.bajdcc.LALR1.interpret.module.net

import com.bajdcc.LALR1.interpret.module.net.ModuleNetServer.Companion.CHANNEL_GROUP
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import io.netty.util.ReferenceCountUtil
import java.util.*

/**
 * 【模块】通讯服务端处理
 *
 * @author bajdcc
 */
class ModuleNetServerHandler(private val msgQueue: Queue<String>) : ChannelInboundHandlerAdapter() {

    override fun channelActive(ctx: ChannelHandlerContext) {
        val ch = ctx.channel()
        CHANNEL_GROUP.add(ch)
        if (CHANNEL_GROUP.size > 0) {
            CHANNEL_GROUP.writeAndFlush(
                    String.format("{ \"origin\": \"%s\", \"addr\": \"%s\", \"type\": \"INFO\", \"content\": \"Hello, client!\" }\r\n",
                            ch.remoteAddress().toString(), ch.localAddress().toString()))
        }
    }

    override fun channelInactive(ctx: ChannelHandlerContext) {
        val ch = ctx.channel()
        msgQueue.add(String.format("{ \"addr\": \"%s\", \"type\": \"INFO\", \"content\": \"Bye, client!\" }\r\n",
                ch.remoteAddress().toString()))
    }

    override fun channelRead(ctx: ChannelHandlerContext, msg: Any) {
        msgQueue.add(msg.toString())
        ReferenceCountUtil.release(msg)
    }

    override fun channelReadComplete(ctx: ChannelHandlerContext) {
        ctx.flush()
    }

    override fun handlerRemoved(ctx: ChannelHandlerContext?) {
        ctx!!.close()
    }

    override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
        cause.printStackTrace()
        CHANNEL_GROUP.remove(ctx.channel())
        ctx.close()
    }
}
